sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Reify.java

```diff
@@ -0,0 +1,140 @@
+      c.outputWithTimestamp(ValueInSingleWindow.of(c.element(), c.timestamp(), window, c.pane()),
+          c.timestamp());
+    }
+  }
+
+  private static class Window<T> extends PTransform<PCollection<T>,
+      PCollection<ValueInSingleWindow<T>>> {
+    @Override
+    public PCollection<ValueInSingleWindow<T>> expand(PCollection<T> input) {
+      return input
+          .apply(ParDo.of(new ValuesInWindowFn<T>()))
+          .setCoder(ValueInSingleWindow.Coder.of(input.getCoder(),
+              input.getWindowingStrategy().getWindowFn().windowCoder()));
+    }
+  }
+
+  private static class WindowInValue<K, V> extends PTransform<PCollection<? extends KV<K, V>>,
+      PCollection<KV<K, ValueInSingleWindow<V>>>> {
+    @Override
+    public PCollection<KV<K, ValueInSingleWindow<V>>> expand(PCollection<? extends KV<K, V>>
+        input) {
+      List<? extends Coder<?>> coderArguments = input.getCoder().getCoderArguments();
+      Coder<K> keyCoder = (Coder<K>) coderArguments.get(0);
+      Coder<V> valueCoder = (Coder<V>) coderArguments.get(1);
+      return input
+          .apply(ParDo.of(new WindowInValueFn<K, V>()))
+          .setCoder(KvCoder.of(keyCoder, ValueInSingleWindow.Coder.of(valueCoder,
+              input.getWindowingStrategy().getWindowFn().windowCoder())));
+    }
+
+    private class WindowInValueFn<K, V> extends DoFn<KV<K, V>, KV<K, ValueInSingleWindow<V>>> {
+      @ProcessElement
+      public void processElement(ProcessContext c, BoundedWindow window) {
+        c.output(KV.of(c.element().getKey(), ValueInSingleWindow.of(c.element()
+            .getValue(), c.timestamp(), window, c.pane())));
+      }
+    }
+  }
+
+  private static class ReifyValueTimestampDoFn<K, V>
+      extends DoFn<KV<K, V>, KV<K, TimestampedValue<V>>> {
+    @ProcessElement
+    public void processElement(ProcessContext context) {
+      context.output(
+          KV.of(
+              context.element().getKey(),
+              TimestampedValue.of(context.element().getValue(), context.timestamp())));
+    }
+  }
+
+  private static class ReifyTimestampDoFn<T> extends DoFn<T, TimestampedValue<T>> {
+    @ProcessElement
+    public void processElement(ProcessContext context) {
+      context.output(TimestampedValue.of(context.element(), context.timestamp()));
+    }
+  }
+
+  Reify() {
+  }
+
+  /**
+   * Create a {@link PTransform} that will output all input {@link KV KVs} with the timestamp inside
+   * the value.
+   */
+  public static <K, V>
+  PTransform<PCollection<? extends KV<K, V>>, PCollection<KV<K, TimestampedValue<V>>>>
+  timestampedValues() {
+    return ParDo.of(new ReifyValueTimestampDoFn<K, V>());
+  }
+
+  /**
+   * Create a {@link PTransform} that will output all inputs wrapped in a {@link TimestampedValue}.
+   */
+  public static <T>
+  PTransform<PCollection<? extends T>, PCollection<TimestampedValue<T>>>
+  timestamps() {
+    return ParDo.of(new ReifyTimestampDoFn<T>());
```

Comment:
Why can this one be a simple "return ParDo.of" but others require their own transform?

Reply:
@jkff `ParDo.of` given that returns a `SingleOutput<InputT, OutputT>` which extends `PTransform<PCollection<? extends InputT>, ...` and given I'm no expert on java generics, I will plan to create more concrete private inner PTransform subclasses in order to get rid of the `? extends` stuff.  This will fix the public signatures and if there's a better solution I don't know about, someone can later simplify the private implementation details.